lighthouse-core/audits/byte-efficiency/large-javascript-libraries.js

```diff
@@ -0,0 +1,167 @@
+
+const Audit = require('../audit.js');
+const i18n = require('../../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that provides detail on large Javascript libraries that are used on the page that have better alternatives. This descriptive title is shown when to users when no known unnecessarily large libraries are detected on the page.*/
+  title: 'Avoids unnecessarily large JavaScript libraries',
+  /** Title of a Lighthouse audit that provides detail on large Javascript libraries that are used on the page that have better alternatives. This descriptive title is shown when to users when some known unnecessarily large libraries are detected on the page.*/
+  failureTitle: 'Replace unnecessarily large JavaScript libraries',
+  /** Description of a Lighthouse audit that tells the user why they should care about the large Javascript libraries that have better alternatives. This is displayed after a user expands the section to see more. No character length limits. */
+  description: 'Large JavaScript libraries can lead to poor performance. ' +
+    'Prefer smaller, functionally equivalent libraries to reduce your bundle size.' +
+    ' [Learn more](https://developers.google.com/web/fundamentals/performance/webpack/decrease-frontend-size#optimize_dependencies).',
+  /** Label for a column in a data table. Entries will be names of large JavaScript libraries that could be replaced. */
+  name: 'Library',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class LargeJavascriptLibraries extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'large-javascript-libraries',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['Stacks'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {LH.Audit.Product}
+   */
+  static audit(artifacts) {
+    /** @type {Array<{original: BundlePhobiaLibrary, suggestions: BundlePhobiaLibrary[]}>} */
+    const libraryPairings = [];
+    const detectedLibs = artifacts.Stacks.filter(stack => stack.detector === 'js');
+
+    const seenLibraries = new Set();
+
+    for (const detectedLib of detectedLibs) {
+      if (!detectedLib.npm || !libStats[detectedLib.npm]) continue;
+      const suggestions = librarySuggestions[detectedLib.npm] || [];
+
+      if (seenLibraries.has(detectedLib.npm)) continue;
+      seenLibraries.add(detectedLib.npm);
+
+      let version = 'latest';
+      if (detectedLib.version && libStats[detectedLib.npm][detectedLib.version]) {
+        version = detectedLib.version;
+      }
+
+      const originalLib = libStats[detectedLib.npm][version];
+      let smallerSuggestions = suggestions.map(suggestion => {
+        if (!libStats[suggestion]) return;
+        if (libStats[suggestion]['latest'].gzip > originalLib.gzip) return;
```

Comment:
based on how we're using the suggestion data, do we ever need size information for anything other than the latest version of our alternatives? seems like we only need per-version size data for our target "large" library

🚲 🏠 might also be clearer what's going on if you always return the mapped objects and then filter on `suggestion => suggestion.gzip < originalLib.gip`